zng/docs/zson.md

```diff
@@ -0,0 +1,628 @@
+```
+|{ {<key>, <value>}, {<key>, <value>}, ... }|
+```
+A type decorator applied to a map can either be one element, referring to a
+map type, or two elements referring to the type of the keys and type of the values.
+
+A map value may be empty.  An empty map value without a type decorator is
+presumed to be an empty map of type (`null`, `null`).
+
+### Type Values
+
+The syntax of a type value or of a decorator expressed as a type value mirrors
+the value syntax.
+
+A primitive type value is the name of the primitive type, i.e., `string`,
+`uint16`, etc.
+
+The syntax of complex type values parallels the syntax of complex values.
+
+A _record type value_ has the form:
+```
+{ <name> : <type>, <name> : <type>, ... }
+```
+where `<type>` is any type value.  The order of the columns is significant,
+e.g., type `{a:int32,b:int32}` is distinct from type `{b:int32,a:int32}`.
+In contrast to schema abstractions in other formats, ZSON has no way to mark
+a field optional as all fields are, in a sense optional: any field can be
+encoded with a null value.  If an instance of a record value omits a value
+by dropping the field altogether rather than using a null, then that record
+value corresponds to a different record type that elides the field in question.
+
+An _array type value_ has the form:
+```
+[ <type> ]
+```
+
+A _set type value_ has the form:
+```
+|[ <type> ]|
+```
+
+A _map type value_ has the form:
+```
+|{ <key-type>, <value-type> }|
+```
+where `<key-type>` is any type value of the keys and `<value-type>` is any
+type value of the values.
+
+A _union type value_ has the form:
+```
+( <type>, <type>, ... )
+```
+where there are at least two types in the list.
+
+An _enum type value_ has two forms.  The simple form is:
+```
+< <identifier>, <identifier>, ... >
+```
+and the complex form is:
+```
+< <identifier>:<value>, <identifier>=<value>, ... >
+```
+In the same form, the underlying enum value is equal to its positional
+index in the list of identifiers as an uint64 type.
+
+## Discussion and Examples
+
+> TBD: Add a range of helpful examples and also some that cover the more obscure corner cases.
+> Example of pico-second timestamp.
+
+### zeek
+```
+{
+    _path: "conn",
+    ts: 2018-03-24T17:15:20.600725Z,
+    uid: "C1zOivgBT6dBmknqk" (bstring),
+    id: {
+        orig_h: 10.47.1.152,
+        orig_p: 49562 (uint16) (=port),
+        resp_h: 23.217.103.245,
+        resp_p: 80 (port)
+    } (=24),
+    proto: "tcp" (=zenum),
+    service: null (bstring),
+    duration: 9.698493s (duration),
+    orig_bytes: 0 (uint64),
+    resp_bytes: 90453565 (uint64),
+    conn_state: "SF" (bstring),
+    local_orig: null (bool),
+    local_resp: null (bool),
+    missed_bytes: 0 (uint64),
+    history: "^dtAttttFf" (bstring),
+    orig_pkts: 57490 (uint64),
+    orig_ip_bytes: 2358856 (uint64),
+    resp_pkts: 123713 (uint64),
+    resp_ip_bytes: 185470730 (uint64),
+    tunnel_parents: null (|[bstring]| (=25))
+} (=26)
```

Comment:
Oh, I guess the internal names are here for illustrative reasons

Reply:
That is output from `-f zson` and when writing a stream you don't know ahead of time which types might be needed.  If you write ZSON by hand or had an automated analysis that looked at the entire text, you could optimize the output by omitting unneeded type defs.

Reply:
Hmm, would be nice to be able to output concise zson... but later!